### PR TITLE
Optional syncing of client settings. 

### PR DESCRIPTION
**Pull request by creesch**
This provides syncing as detailed in #1825. Some demonstrations

- [General behavior of syncing and unchecking the box.](http://creesch.com/dump/vid/vid_5a3432fc0f6dc.mp4)
- [Enabling sync for the very first time](https://www.creesch.com/dump/vid/vid_5a3692e0339c4.mp4)

## Note on always syncing some settings. 

This PR includes the ability to make some settings mandatory to sync regardless of syncing in general being on or not. This works fine, however I didn't include any settings as migration is a consideration. The ability was requested in relation to custom highlights, not working for people with push enabled.  So when people have no push enabled then it isn't broken to them.  Meaning that it is likely that some people have explicitly set it differently in different clients, which means that a migration strategy is needed. 

Because the goal of this PR is to implement syncing in general first I decided to not turn mandatory sync yet. I left the ability in as the goal of this PR is to have a basis to work on from here on out. 

## options.js refactoring 
In addition this PR also reworks `client/js/options.js` to be more reliable, easier to understand and generally more robust. Functionality that was all randomly called on has been divided in separate functions: 

- Initialization 
- Processing 
- Saving 
- Applying 

Meaning that settings are now actually only saved and applied when things actually change. Before it could easily happen that when nothing was actually changed in settings there still would be multiple write actions and attempts at applying settings to the client where it was not needed. 

It also changes the `module.exports` object to separate user settings from other information and functions, potentially preventing outside modules from accidentally overwriting settings. 

Making options.js robust enough to be able to handle synced settings was the biggest part of the work that needed to be done.

### CODE REVIEW

File: client/js/socket-events/init.js [hunk 1]
```diff
@@ -4,12 +4,18 @@ const $ = require("jquery");
...
 const storage = require("../localStorage");
 const utils = require("../utils");
 
 socket.on("init", function(data) {
+	if (options.initialized) {
+		return;
+	}
+	options.initialize(data.clientSettings ? data.clientSettings : "");
+	webpush.initialize();
 	$("#loading-page-message, #connection-error").text("Rendering…");
```
Comment (xPaw) [inline client/js/socket-events/init.js]: Put rendering above initializes.

File: client/js/socket-events/init.js [hunk 2]
```diff
@@ -4,12 +4,18 @@ const $ = require("jquery");
...
 const socket = require("../socket");
 const render = require("../render");
+const options = require("../options");
 const webpush = require("../webpush");
 const sidebar = $("#sidebar");
 const storage = require("../localStorage");
 const utils = require("../utils");
 
 socket.on("init", function(data) {
+	if (options.initialized) {
```
Comment (xPaw) [inline client/js/socket-events/init.js]: I'm quite certain doing this breaks reconnections.

File: src/server.js [hunk 3]
```diff
@@ -388,6 +388,42 @@ function initializeClient(socket, client, token, lastMessage) {
 
 	socket.on("sessions:get", sendSessionList);
 
+	socket.on("setting:set", (newSetting) => {
+		// Older user configs will not have the clientSettings property.
+		if (!client.config.hasOwnProperty("clientSettings")) {
+			client.config.clientSettings = {};
+		}
+		// We do not need to do write operations and emit events if nothing changed.
+		if (!client.config.clientSettings.hasOwnProperty(newSetting.name) || client.config.clientSettings[newSetting.name] !== newSetting.value) {
```
Comment (xPaw) [inline src/server.js]: `hasOwnProperty` is unnecessary here.

File: src/server.js [hunk 4]
```diff
@@ -388,6 +388,42 @@ function initializeClient(socket, client, token, lastMessage) {
...
+			});
+		}
+	});
+
+	socket.on("settings:get", () => {
+		if (!client.config.hasOwnProperty("clientSettings")) {
+			socket.emit("setting:empty");
+		} else {
+			const clientSettings = client.config.clientSettings;
+			for (const setting in clientSettings) {
```
Comment (xPaw) [inline src/server.js]: Seems very wasteful to emit separate packets for each setting. You're better off doing one emit and handling the loop on the client.

File: client/js/socket-events/init.js (hunk 2 above)
Comment (creesch) [reply client/js/socket-events/init.js]: This is already done in a very similar way currently, are you sure about that?

Comment (xPaw) [reply client/js/socket-events/init.js]: Your change blocks future `init` calls, which happen after reconnecting. On master it blocks `configuration` which is fine.

File: src/server.js (hunk 4 above)
Comment (creesch) [reply src/server.js]: Fair enough, I'll rework that.

File: src/server.js [hunk 5]
```diff
@@ -388,6 +388,42 @@ function initializeClient(socket, client, token, lastMessage) {
...
+			client.emit("setting:new", {
+				name: newSetting.name,
+				value: newSetting.value,
+			});
+		}
+	});
+
+	socket.on("settings:get", () => {
+		if (!client.config.hasOwnProperty("clientSettings")) {
+			socket.emit("setting:empty");
```
[the remark this answers is not in the document]
Comment (creesch) [reply src/server.js]: Yup, though I'd think it is better to still make it two events 

  - `setting:new` which is already a thing for when individual settings are saved and synced and we know the setting is new.. 
  - `settings:all` for the entire package where the client can figure out looping etc if needed. 

Makes things more explicit and also reduces the amount of looping that needs to be done in general for figuring out what needs to be changed.

[the remark this answers is not in the document]
Comment (xPaw) [reply src/server.js]: If both of these events will handle settings in the same way (except for the loop), then I don't see a reason to do separate events.

File: src/server.js [hunk 6]
```diff
@@ -388,6 +388,34 @@ function initializeClient(socket, client, token, lastMessage) {
...
+			client.emit("setting:new", {
+				name: newSetting.name,
+				value: newSetting.value,
+			});
+		}
+	});
+
+	socket.on("settings:get", () => {
+		if (!client.config.hasOwnProperty("clientSettings")) {
+			socket.emit("setting:all", {});
```
Comment (xPaw) [inline src/server.js]: You forgot a return.

Comment (creesch) [reply src/server.js]: Actually forgot the else, but return works as well. Fixing that now.

File: client/views/windows/settings.tpl [hunk 7]
```diff
@@ -5,6 +5,23 @@
...
+			<h2>
+				Settings synchronisation
+			</h2>
+			<p class="opt">
+				<input type="checkbox" name="syncSettings">
+				Synchronize settings with other clients. <br>
+				<strong>Warning</strong> Checking this box will override current client settings.
+			<p>
+			<p id="master-sync" class="opt">
+				No settings synced yet, sync settings from this client as base set?
```
Comment (YaManicKill) [inline client/views/windows/settings.tpl]: I'm not convinced on the value of this?

Under what circumstances would someone want to sync when there are no server settings and *not* set these as the ones currently synced.

File: client/js/socket-events/configuration.js [hunk 8]
```diff
@@ -7,10 +7,9 @@ const options = require("../options");
 const webpush = require("../webpush");
 
 socket.on("configuration", function(data) {
-	if (!options.initialize) {
+	if (options.initialized) {
```
Comment (YaManicKill) [inline client/js/socket-events/configuration.js]: Hmmm, why this change?

File: client/js/options.js [hunk 9]
```diff
@@ -22,156 +35,263 @@ const options = {
...
-	userStyles: userStyles.text(),
+	userStyles: "",
 };
-let userOptions = JSON.parse(storage.get("settings")) || {};
 
-for (const key in options) {
-	if (userOptions[key] !== undefined) {
-		options[key] = userOptions[key];
+const noSync = ["syncSettings"];
+// Reserved for things like "highlights".
```
Comment (YaManicKill) [inline client/js/options.js]: Nitpick: it's not clear which this is referring to. I initially thought it was noSync it was talking about.

File: client/js/socket-events/configuration.js (hunk 8 above)
Comment (creesch) [reply client/js/socket-events/configuration.js]: Before it wasn't a boolean and checked something else.

File: client/views/windows/settings.tpl (hunk 7 above)
Comment (creesch) [reply client/views/windows/settings.tpl]: The first time someone does enable syncing there is nothing on the server yet. We could automatically sync the settings of that client but they might not have intended that. [See also this video](https://www.creesch.com/dump/vid/vid_5a3692e0339c4.mp4) for how it functions. 
This will basically give them the opportunity to go "Oh crap, I didn't want to sync this client but my other one". 

Basically done from the same reasoning as how I approached the advanced settings classification the other day :)

File: client/js/options.js (hunk 9 above)
Comment (creesch) [reply client/js/options.js]: The comment you mean?

File: client/js/socket-events/configuration.js (hunk 8 above)
Comment (creesch) [reply client/js/socket-events/configuration.js]: To be a bit more clear, before it set `module.exports.initialize` to `null` which works (but a bit overkill) but also resulted in the negative check for initialize which is slightly confusing as it is more of a `true` situation as options are in fact initialized. 

tl;dr readability of code mostly.

File: client/views/windows/settings.tpl (hunk 7 above)
Comment (YaManicKill) [reply client/views/windows/settings.tpl]: I don't think you've convinced me of this one. I think it's a bit of a weird thing, but ... I'll let others say what they think. The rest I really like, I'm just not convinced about this 1 button press for when there are no synced settings and never again.

File: client/js/socket-events/configuration.js (hunk 8 above)
Comment (YaManicKill) [reply client/js/socket-events/configuration.js]: Aha, I see

File: client/js/options.js (hunk 9 above)
Comment (YaManicKill) [reply client/js/options.js]: Yeah

File: client/views/windows/settings.tpl (hunk 7 above)
Comment (creesch) [reply client/views/windows/settings.tpl]: As per the discussion on irc I'll remove the manual step and will instead display something like 

>  "No settings have been synced before. Enabling this will make the settings of this client the base settings."

File: client/js/options.js [hunk 10]
```diff
@@ -22,156 +36,269 @@ const options = {
 	notifyAllMessages: false,
 	showSeconds: false,
 	statusMessages: "condensed",
-	theme: $("#theme").attr("href").replace(/^themes\/(.*).css$/, "$1"), // Extracts default theme name, set on the server configuration
+	theme: $theme.attr("href").replace(/^themes\/(.*).css$/, "$1"), // Extracts default theme name, set on the server configuration
```
Comment (xPaw) [inline client/js/options.js]: Can you change all `.attr` to `.prop` please, this was done on master.

Comment (creesch) [reply client/js/options.js]: Yeah sure. 

Out of curiosity, why has it been changed?

Comment (xPaw) [reply client/js/options.js]: https://github.com/thelounge/lounge/pull/2015#discussion_r164976035

Comment (creesch) [reply client/js/options.js]: Changed it. Also rebased and squashed it.

**Comment by McInkay** [comment]
I think this PR should be merged (if we want this) reasonably soon. There are a lot of extra pieces of functionality I want to implement that will change code here, and the improvements to `options.js` will be necessary for them.

As a side-note, this can easily be added as an advanced feature if #1860 is merged. It can be done in either PR depending on which is done before, but neither will conflict much with the other, as that one is almost entirely in html (almost).

**Comment by creesch** [comment]
Yup, this is probably one in the advanced section.

File: src/server.js [hunk 11]
```diff
@@ -432,6 +432,35 @@ function initializeClient(socket, client, token, lastMessage) {
 
 	socket.on("sessions:get", sendSessionList);
 
+	socket.on("setting:set", (newSetting) => {
```
Comment (xPaw) [inline src/server.js]: Don't bind these setting events in public mode.

**Comment by creesch** [comment]
@xPaw made the changes to public mode, not sure why Travis is failing on node9. Seems like a timeout and not an actually failed test.

File: client/js/options.js [hunk 12]
```diff
@@ -2,15 +2,29 @@
...
+const $userStyles = $("#user-specified-css");
 
-// Default options
-const options = {
+const noCSSparamReg = /[?&]nocss/;
+
+// Not yet available at this point but used in various functionaly.
+// Will be assigned when `initialize` is called.
+let $syncWarningOverride;
+let $syncWarningBase;
```
Comment (MaxLeiter) [inline client/js/options.js]: I don’t think we prefix element objects anywhere else with $

Comment (creesch) [reply client/js/options.js]: It is a good habit to do it with jqueury objects, makes it easier to read code.

File: client/views/windows/settings.tpl [hunk 13]
```diff
@@ -5,6 +5,19 @@
 	<h1 class="title">Settings</h1>
 
 	<div class="row">
+		{{#unless public}}
+		<div class="col-sm-12">
+			<h2>
+				Settings synchronisation
+			</h2>
+			<p class="opt">
+				<input type="checkbox" name="syncSettings">
```
Comment (xPaw) [inline client/views/windows/settings.tpl]: Make this a correct `<label>` so the text is also clickable. (look at other checkboxes)

Don't include warnings as part of the label though.

Comment (creesch) [reply client/views/windows/settings.tpl]: I explicitly didn't do that as that creates a bunch of issues with settings styling and how things are aligned and you really don't want the text to be part of the label as you don't want to accidentally activate this. 

Also just having the input element in a label doesn't make much sense to me?

Comment (creesch) [reply client/views/windows/settings.tpl]: Done

File: client/views/windows/settings.tpl [hunk 14]
```diff
@@ -5,6 +5,19 @@
...
+		<div class="col-sm-12">
+			<h2>
+				Settings synchronisation
+			</h2>
+			<label class="opt">
+				<input type="checkbox" name="syncSettings">
+				Synchronize settings with other clients.
+			</label>
+			<span class="sync-warning-override"><strong>Warning</strong> Checking this box will override the settings of this client with those stored on the server.</span>
+			<span class="sync-warning-base"><strong>Warning</strong> No settings have been synced before. Enabling this will sync all settings of this client as the base for other clients.</span>
```
Comment (astorije) [inline client/views/windows/settings.tpl]: Could use `p` elements instead of `span`s? That will fix the extra font size there is right now.

File: src/server.js [hunk 15]
```diff
@@ -441,6 +441,44 @@ function initializeClient(socket, client, token, lastMessage) {
...
+					value: newSetting.value,
+				});
+
+				client.manager.updateUser(client.name, {
+					clientSettings: client.config.clientSettings,
+				});
+			}
+		});
+
+		socket.on("settings:get", () => {
```
Comment (astorije) [inline src/server.js]: In terms of socket events, there are both `setting` (including `setting:all`) and `settings` like this guy. Do we want to use `settings` everywhere instead?

File: client/views/windows/settings.tpl (hunk 14 above)
Comment (creesch) [reply client/views/windows/settings.tpl]: I actually liked the warning being slightly bigger, but yeah it can also be changed if you prefer so.

Comment (astorije) [reply client/views/windows/settings.tpl]: That would be great, thanks :)

**Comment by astorije** [comment]
Hey @creesch, I quickly looked at the code and tested, and didn't notice any bugs. For the full, detailed review, I'll defer to @xPaw and @McInkay, who have been very involved in this so far, so I think I would not bring anything else useful to the table.
Feel free to ping me if you need a hand with, like, final UI or something like this, but I don't think this will massively change.

File: src/server.js (hunk 15 above)
Comment (creesch) [reply src/server.js]: Yeah that makes sense, initially I had the singular and plural because one was going to pass multiple settings and the other not. That changed along the way so it makes sense to have consistent naming there.

**Comment by creesch** [comment]
@astorije made those changes. 

At the request of @xPaw also added an extra warning note that looks like this

image

As a bonus, that can in the future also be used for other things.

**Comment by astorije** [comment]
Hey @creesch, we have free sticker packs for our contributors now!
If you're interested, go fill the form at https://goo.gl/forms/f5usqAEp5DWoeXC92 🙂
